[backend] request handler that returns green commits

XilaiZhang opened the pull request:
An attempt to address https://github.com/flutter/flutter/issues/99869, and to create a request handler that returns commit shas of green commits.

A green commit satisfies rules:
1. A commit had all its tests run (at least those that are not in bringup)
2. All those blocking tasks were green

Currently I am returning commits that have all its stages with status [TaskSucceeded]. My understanding of rule 1 is that commit doesn't have [Task.statusInprogress]. And I wasn't very sure what "blocking" tasks are, and assumed that we would fulfill rule 2 if all tasks succeed.

My understanding of the rules could be very off, and I am more than happy to rework.

CaseyHillers commented on app_dart/lib/src/request_handlers/get_green_commits.dart:
```diff
@@ -0,0 +1,57 @@
+          branch: branch,
+          slug: slug,
+        )
+        .where((CommitStatus status) => status.stages.every((Stage s) => s.taskStatus == Task.statusSucceeded))
```
As is, we'll get 0 commits from this.

You need to filter out the tasks that are marked in bringup

CaseyHillers commented on app_dart/bin/server.dart:
```diff
@@ -216,6 +216,33 @@ Future<void> main() async {
         delegate: GetStatus(config),
       ),
 
+      /// Returns a list of the commit shas that had green runs.
```
This doc should be in the class.

See https://github.com/flutter/flutter/issues/97675

CaseyHillers commented on app_dart/bin/server.dart:
```diff
@@ -216,6 +216,33 @@ Future<void> main() async {
+      ///
+      /// GET: /api/public/get-green-commits?repo=$repo
+      ///
+      /// Response: Status: 200 OK
```
I would avoid hard coding a response here, but instead say it's a List<String> of the commit shas

XilaiZhang replied:
umm Maybe my understanding is not correct, but I think the get() methods in request handlers override the Future<Body> get() function, and returns a Body type, which from my understanding represents a HTTP response body. Correct me if I am wrong, I think if we use a HTTP get request and it succeeds with a response body, my understanding is that we are not hard coding it?

CaseyHillers replied:
You're hard coding an example response in the documentation. Prefer giving the type returned instead as it's more concise and obvious to update when we change the logic here

XilaiZhang replied:
I see. My bad, I misunderstood it as my response was hard coded instead of generated.

CaseyHillers commented on app_dart/bin/server.dart:
```diff
@@ -216,6 +216,33 @@ Future<void> main() async {
+      /// }
+      ///
+
+      '/api/public/get-green-commits': CacheRequestHandler<Body>(
```
This isn't going to get called very often, so we can remove the cache wrapper

CaseyHillers commented on app_dart/lib/src/request_handlers/get_green_commits.dart:
```diff
@@ -0,0 +1,57 @@
+        .toList();
+
+    return Body.forJson(<String, List<String?>>{
+      'greenCommits': greenCommits,
```
Why return a Map? You can just return a List<String> which is easier to use

XilaiZhang replied:
umm Maybe my understanding is not correct, but I think the get() methods in request handlers have to be a valid override of the `Future<Body> get()` function, which requires a `Future<Body>` as a return type. And to create a Body type, the three valid ways are through `Body.forString`, `Body.forStream` and `Body.forJson`, which requires input type as String, Stream<Uint8List?> and dynamic(json) respectively. I am not sure how to cast List<String> into a String or Stream<Uint8List?> type, and so I ended up making it a valid json type (a map, sort of), and then feed it into Body.forJson to create a Body type. umm Would there be a valid creation of Body type if we use List<String> type?

CaseyHillers replied:
List<String> is valid json. You can use Body.fromJson(listStringVar)

XilaiZhang replied:
Thanks for the hint! I have now casted Stream.single to List<dynamic> and then to List<String> to make it work for List type.

CaseyHillers commented on app_dart/test/request_handlers/get_green_commits_test.dart:
```diff
@@ -0,0 +1,156 @@
+        datastoreProvider: (DatastoreDB db) => DatastoreService(config.db, 5),
+        buildStatusProvider: (_) => buildStatusService,
+      );
+      commit1 = Commit(
```
Take advantage of the `entity_generators.dart` class. It allows you to do `generateCommit` and `generateTask` to remove the boiler plate

CaseyHillers commented on app_dart/test/request_handlers/get_green_commits_test.dart:
```diff
@@ -0,0 +1,156 @@
+        status: Task.statusSucceeded,
+      );
+
+      stage1 = Stage('cocoon', commit1, [task1, task2], Task.statusSucceeded);
```
Can we avoid using stages? We're looking to remove them soon

XilaiZhang replied:
Sure. I have updated validation logic in request handler to be based on individual [Task] instead of [stage]. umm In tests, I might be wrong, but I think given the current topology of accessing tasks from commitStatus: [commitStatus] -> List<stage> -> List<task> , it would be necessary to create [stage] variables as intermediate bridges to retrieve the underlying [Task] variables?

CaseyHillers replied:
The tests are fine then

XilaiZhang commented:
Thanks a lot for the instructions! I resolved part of them and left comments for the rest. Would be great if you can take a look at the comments and provide some insights, thank you!

XilaiZhang commented:
Thanks a lot for the reply! I have now removed example response and used List type

CaseyHillers commented on app_dart/lib/src/request_handlers/get_green_commits.dart:
```diff
@@ -0,0 +1,74 @@
+import '../service/config.dart';
+import '../service/datastore.dart';
+
+/// Returns a list of the commit shas that had green runs.
```
Take advantage of the dart docs format

```suggestion
/// Returns [List<String>] of commit shas that had all passing tests.
```

CaseyHillers commented on app_dart/lib/src/request_handlers/get_green_commits.dart:
```diff
@@ -0,0 +1,74 @@
+
+/// Returns a list of the commit shas that had green runs.
+///
+/// A green commit is used to help the release tooling find commits Flutter infrastructure has validated. The rules are:
```
The terminology here should use `[CommitStatus]` as it's an aggregate of the `[List<Task>]` for a `[Commit]`

CaseyHillers commented on app_dart/lib/src/request_handlers/get_green_commits.dart:
```diff
@@ -0,0 +1,74 @@
+/// A green commit is used to help the release tooling find commits Flutter infrastructure has validated. The rules are:
+/// 1. A commit had all its tests run (at least those that are not in bringup)
+/// 2. All those blocking tasks were green
+/// Green commit shas are returned in the order of commit timestamp.
```
What order?

CaseyHillers commented on app_dart/test/request_handlers/get_green_commits_test.dart:
```diff
@@ -0,0 +1,126 @@
+    late RequestHandlerTester tester;
+    late GetGreenCommits handler;
+
+    late Commit commit1;
```
Since this test doesn't change data, we should make these final instead of late

CaseyHillers commented on app_dart/test/request_handlers/get_green_commits_test.dart:
```diff
@@ -0,0 +1,126 @@
+import '../src/utilities/entity_generators.dart';
+
+void main() {
+  group('GetStatus', () {
```
nit: GetGreenCommits

XilaiZhang commented:
Thanks for the reply! I have moved late variables out of `Setup()` and make them final. I have added explanation on the order of returned shas, replaced the wording of `green` with `passing all tests`, and added more brackets to detail the data types.

CaseyHillers commented on app_dart/lib/src/request_handlers/get_green_commits.dart:
```diff
@@ -0,0 +1,78 @@
+    final DatastoreService datastore = datastoreProvider(config.db);
+    final BuildStatusService buildStatusService = buildStatusProvider(datastore);
+    final int commitNumber = config.commitNumber;
+    final int lastCommitTimestamp = DateTime.now().millisecondsSinceEpoch;
```
Is `lastCommitTimestamp` needed?

CaseyHillers commented on app_dart/test/request_handlers/get_green_commits_test.dart:
```diff
@@ -0,0 +1,113 @@
+      expect(result, isEmpty);
+    });
+
+    test(
```
It's best practice to keep your tests small, and only validate a specific aspect. For this, you can tell this test is doing too much by having "and" and "but" mulitple times in the description. In these cases, either split into smaller tests or wrap in ` group`

CaseyHillers commented on app_dart/lib/src/request_handlers/get_green_commits.dart:
```diff
@@ -54,12 +53,11 @@ class GetGreenCommits extends RequestHandler<Body> {
         .retrieveCommitStatus(
           limit: commitNumber,
-          timestamp: lastCommitTimestamp,
+          timestamp: DateTime.now().millisecondsSinceEpoch,
```
You don't need to pass timestamp (by default it'll be the current time).

The reason we want to avoid this is using DateTime.now() will cause flakiness in tests

CaseyHillers commented on app_dart/test/request_handlers/get_green_commits_test.dart:
```diff
@@ -86,11 +84,48 @@ void main() {
     });
 
-    test('Also select green commits that include failed tasks but have bringup: true label', () async {
+    test('Also return commits with both flaky and succeeded tasks', () async {
```
nit: rm Also

CaseyHillers reviewed (commented):
LGTM. Just some last comments

XilaiZhang commented:
thanks for the previous approval! I added in a few lines to clean up the side effect of removing timestamp, would like to get them verified before I merge, thank you!

CaseyHillers commented:
![flutter-lgtm](https://user-images.githubusercontent.com/19588/120398463-2c3b7b00-c2ef-11eb-88f0-7eee0d3c01b0.png)
